src/pages/data-first-and-data-last-a-comparison.md

```diff
@@ -0,0 +1,474 @@
+
+The compiler now can infer that the `u` expression in the callback parameter has type `User.t`, and so when it sees the `u.age` expression on the right side, it can be 100% sure where it comes from, and make sure the body of the callback is valid.
+
+### The pipe first operator `->`
+
+In the same vein as the pipe operator `|>`, BuckleScript introduced a [pipe operator](https://bucklescript.github.io/docs/en/pipe-first) that is similar, but instead passes the resulting value of the expression in the left side as the _first_ parameter of the one on the right.
+
+So, for example:
+
+```reason
+let filtered = list -> Belt.List.filter(a => a > 1);
+```
+
+is equivalent to:
+
+```reason
+let filtered = Belt.List.filter(list, a => a > 1);
+```
+
+Another important difference from the `|>` operator is that `->` is not an infix operator, just syntactic sugar, so it is really as if you were writing the second form instead of the first from the compiler perspective. With the traditional pipe `|>` it is interpreted like applying a function.
+
+## Advantages and disadvantages of data-first
+
+As we have seen, data-first remains more closely to how type inference works. We also saw how this:
+
+- Helps the compiler infer types in functions that take callbacks as parameters, without having to manually add type annotations.
+- Makes error messages simpler.
+
+This might not seem like a very big deal on its own, but it has a lot of impact down the line that affects the resulting developer experience.
+
+### Better IDE integration
+
+A direct consequence of the data-first approach that benefits from the compiler having more information is that we get more help from editor extensions when writing our functions.
+
+In the example above, as we are writing it:
+
+```reason
+let ages = Belt.List.map(User.admins, u => u.
+                                             ◻️ age
+                                             ◻️ name
+```
+
+The editor extensions, as they rely on the information provided by compiler, can know that `u` is a value of the record type `User.t`, and can provide autocompletion for the fields in it. Very helpful!
+
+The advantages of the data-first approach when it comes to editor integration is something that language designers with a vast experience like Anders Hejlsberg —creator and lead architect of TypeScript— [have pointed out in the past](https://github.com/Microsoft/TypeScript/issues/15680#issuecomment-307571917).
+
+### Intuitive design for functions with multiple params
+
+One of the downsides of the data-last approach is that sometimes it makes harder to understand what a function with two operands is doing.
+
+For example, the `String.concat` function in OCaml standard library:
+
+```reason
+let foo = String.concat("a", ["b", "c"])
+```
+
+Because we are used to left-to-right reading —like the inference engine— we could guess the resulting value of `foo` is the string `"abc"`, but it's actually `"bca"` because the API is data-last.
```

Comment:
Actually, the result is "bac" because the first parameter is the separator string that is placed between the strings in the second parameter's string list. Anyway, strictly speaking, I wouldn't say that this particular function is data-last, because if that were the case we would expect a `String.t` i.e. `string` as the last parameter. The functions in this module don't really adhere to data-last, e.g. `iter` is data-last but `contains` is data-first.